vip-jetpack/connection-pilot/class-jetpack-connection-status-check.php

```diff
@@ -0,0 +1,200 @@
+	 * Populates the class properties and sets a fallback error.
+	 */
+	public function __construct() {
+		$this->healthcheck_option = get_option( self::HEALTHCHECK_OPTION_NAME );
+		$this->site_url           = get_site_url();
+		$this->connection_error   = new WP_Error( 'jp-cxn-pilot-default-error', 'The Jetpack connection is in an unknown state.' );
+	}
+
+	/**
+	 * The main method, used to start the status check.
+	 *
+	 * Runs connection checks and handles any problems.
+	 */
+	public function launch() {
+		$is_connected = WPCOM_VIP_Jetpack_Connection_Controls::jetpack_is_connected();
+
+		if ( true === $is_connected ) {
+			// Everything checks out. Update the healthcheck option and move on.
+			return update_option( self::HEALTHCHECK_OPTION_NAME, array(
+				'site_url'         => $this->site_url,
+				'cache_site_id'    => (int) Jetpack_Options::get_option( 'id' ),
+				'last_healthcheck' => time(),
+			), false );
+		}
+
+		if ( is_wp_error( $is_connected ) ) {
+			$this->connection_error = $is_connected;
+		}
+
+		// The connection check failed, let's handle the problem the best we can.
+		$this->handle_connection_issues();
+	}
+
+	/**
+	 * Handle connection issues. (Re)connect when possible, else add a notification.
+	 */
+	private function handle_connection_issues() {
+		if ( in_array( $this->connection_error->get_error_code(), array( 'jp-cxn-pilot-missing-constants', 'jp-cxn-pilot-development-mode' ), true ) ) {
+			return $this->notify_pilot( 'Jetpack cannot currently be connected on this site.' );
+		}
+
+		// 1) It is connected but not under the right account.
+		if ( 'jp-cxn-pilot-not-vip-owned' === $this->connection_error->get_error_code() ) {
+			return $this->notify_pilot( 'Jetpack is connected to a non-VIP account.' );
+		}
+
+		// 2) Check the last healthcheck to see if the URLs match.
+		if ( ! empty( $this->healthcheck_option['site_url'] ) ) {
+			return $this->handle_disconnects_using_last_healthcheck();
+		}
+
+		// 3) The healthcheck option doesn’t exist. Either it's a new site, or an unknown connection error.
+		if ( $this->is_default_vip_domain() || $this->is_new_multisite_site() ) {
+			return $this->handle_new_sites();
+		} else {
+			return $this->notify_pilot( 'Jetpack is disconnected.' );
+		}
+	}
+
+	/**
+	 * Connection issue handler for sites that have previous healthcheck data.
+	 *
+	 * Will attempt to automatically reconnect if the URLs are the same.
+	 */
+	private function handle_disconnects_using_last_healthcheck() {
+		if ( $this->healthcheck_option['site_url'] === $this->site_url ) {
+			$reconnection_attempt = $this->connect_site();
+			return $this->notify_pilot( $reconnection_attempt );
+		}
+
+		return $this->notify_pilot( 'Jetpack is disconnected, and it appears the domain has changed.' );
+	}
+
+	/**
+	 * Connection issue handler for new sites.
+	 *
+	 * This includes either a default go-vip domain
+	 * or a recently created multi-site site.
+	 */
+	private function handle_new_sites() {
```

Comment:
Looks like this doesn't actually attempt to connect the site; is that intentional?

Reply:
It also might be better to just separate into two functions since w're repeating the `if` logic here and where it's called.

Reply:
Yes, this is intentional at the moment. The scope for the initial rollout was to only auto-reconnect in the one scenario. Connecting new sites was out of scope, but definitely something that I want to do, just perhaps needs a little more discussion first. That said, technically speaking, it's all ready to be done with like a single line change.

The conditional won't really need to exist inside the `handle_new_sites()` method once we are auto-connecting since both will be handled the same way. I've just left it there as it made it nice while testing to ensure I'm getting the right message.